### ISSUE TITLE
Salt Virt runner minion seeding failing

### ISSUE THREAD

**Issue by q1x** [saltstack/salt#43928]
### Description of Issue/Question
When using the salt virt runner to deploy a new vm to a KVM host, like so:

`salt-run virt.init mynewminion 1 1024 /var/lib/libvirt/images/image.qcow2 nic=nicprofile host='kvmhost'`

The preseeding of the minion fails with the following exception:

```
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1468, in _thread_return
        return_data = executor.execute()
      File "/usr/lib/python2.7/site-packages/salt/executors/direct_call.py", line 28, in execute
        return self.func(*self.args, **self.kwargs)
      File "/usr/lib/python2.7/site-packages/salt/modules/virt.py", line 691, in init
        __salt__[seed_cmd](
      File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1113, in __getitem__
        func = super(LazyLoader, self).__getitem__(item)
      File "/usr/lib/python2.7/site-packages/salt/utils/lazy.py", line 96, in __getitem__
        if self._load(key):
      File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1578, in _load
        raise KeyError
    KeyError
```
The debug log of the minion shows the following lines:
```
[DEBUG   ] Generating VM XML
[DEBUG   ] Seed command is False
[WARNING ] The minion function caused an exception
```

If we however use the following salt command, seeding succeed but the minion-key is not automatically accepted on the master.

`salt 'kvmhost' virt.init mynewminion 1 1024 image=/var/lib/libvirt/images/image.qcow2 nic=nicprofile`

The debug log shows the following lines:
```
[DEBUG   ] Generating VM XML
[DEBUG   ] Seed command is seed.apply
[DEBUG   ] LazyLoaded seed.apply
```
After which guestmount is used to do the preseeding.

It seems the Salt runner needs to be told to use `seed.apply` as the seed command.

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

Salt master:
```
Salt Version:
           Salt: 2017.7.1

Dependency Versions:
           cffi: 1.10.0
       cherrypy: 5.6.0
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.24.6
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: 0.24.2
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.3.1611 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-514.21.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core
```

KVM host:
```
Salt Version:
           Salt: 2017.7.1

Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: Not Installed
         Python: 2.7.5 (default, Aug  4 2017, 00:39:18)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.4.1708 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-693.2.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core
```

**Comment by q1x** [comment]
To be clear, the VM is started in both scenarios, the fist one is not seeded, the second is.
However, the first scenario accepts the minion key on the master, the second does not.

Ideally, I would have a situation where both the minion key is accepted on the master and the minion is preseeded :)

**Comment by gtmanfred** [comment]
Unfortunately we do not run a lot of tests on salt-virt.  Would you be able to provide me with some salt states to setup an environment that is able to replicate this issue?

Thanks,
Daniel

**Comment by q1x** [comment]
I've basically followed this tutorial to get things up and running:

https://docs.saltstack.com/en/latest/topics/tutorials/cloud_controller.html#cloud-controller

The only differences here are that I'm using a custom KVM qcow2 image (located on the hypervisor filesystem) which uses our internal yum repos and already has the Salt minion installed in the image.

Also, I'm not using the capabilities of `virt.keys` (as per tutorial) but I'm using certs on the KVM services signed by our internal CA instead. This all works as the `virt.query` function is working as expected.

**Comment by gtmanfred** [comment]
Alright, I will try to get to setting this up and replicating it early next
week.

**Comment by q1x** [comment]
Thank you, let me know if you need more info!

**Comment by kaaelhaa** [comment]
I too just experienced this.

It seems the parameters passed to the virt module from the virt runner is out of order (specifically, the virt runner passes the `enable_vnc` parameter to the virt modules `seed_cmd` parameter).

I'll try to cook up a PR with a fix within a few days.

**Comment by gtmanfred** [comment]
Awesome, thanks!

Sorry i forgot about this one.

Thanks,
Daniel

**Comment by Zauberfisch** [comment]
just to bump this issue (I can confirm it's still present in `2018.3.2 (Oxygen)`):

I experienced the same issue while following the [SALT AS A CLOUD CONTROLLER](https://docs.saltstack.com/en/latest/topics/tutorials/cloud_controller.html) tutorial.

What threw me off, was that it basically failed without error message and still reported that the VM has been created, yet `virsh list --all` on the hypervisor returned an empty list.    
    
normal output of `salt-run -virt.init ubuntu16 2 2048 salt://hypervisor/images/ubuntu-18.04-server-cloudimg-amd64.img`:
```
event:
    ----------
    _stamp:
        2018-07-24T11:08:41.336613
    message:
        Searching for hosts
suffix:
    progress
event:
    ----------
    _stamp:
        2018-07-24T11:08:57.263275
    message:
        Minion will be preseeded
suffix:
    progress
event:
    ----------
    _stamp:
        2018-07-24T11:08:57.996350
    message:
        Creating VM ubuntu16 on host hypervisor.local
suffix:
    progress
event:
    ----------
    _stamp:
        2018-07-24T11:09:16.915566
    message:
        VM ubuntu16 initialized on host hypervisor.local
suffix:
    progress
good

```    

though `salt-run -l all virt.init ubuntu17 2 2048 salt://hypervisor/images/ubuntu-18.04-server-cloudimg-amd64.img` shows an exception being thrown:
```
[...]
[TRACE   ] get_event() received = {u'tag': 'salt/job/20180724062031469479/ret/hypervisor.local', u'data': {u'fun_args': [u'ubuntu12', 2, 2048, u'salt://hypervisor/images/ubuntu-18.04-server-cloudimg-amd64.img', u'default', u'kvm', True, u'default', u'base', True, True, u'-----BEGIN PUBLIC KEY-----[...]-----END PUBLIC KEY-----\n', u'-----BEGIN RSA PRIVATE KEY-----[...]-----END RSA PRIVATE KEY-----\n', False], u'jid': u'20180724062031469479', u'return': u'The minion function caused an exception: Traceback (most recent call last):\n  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1606, in _thread_return\n    return_data = minion_instance.executors[fname](opts, data, func, args, kwargs)\n  File "/usr/lib/python2.7/site-packages/salt/executors/direct_call.py", line 12, in execute\n    return func(*args, **kwargs)\n  File "/usr/lib/python2.7/site-packages/salt/modules/virt.py", line 767, in init\n    __salt__[seed_cmd](\n  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1143, in __getitem__\n    func = super(LazyLoader, self).__getitem__(item)\n  File "/usr/lib/python2.7/site-packages/salt/utils/lazy.py", line 99, in __getitem__\n    if self._load(key):\n  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1621, in _load\n    raise KeyError(\'The key must be a string.\', key)\nKeyError: (u"The key must be a string.")\n', u'success': False, u'cmd': u'_return', u'_stamp': u'2018-07-24T06:20:46.926662', u'fun': u'virt.init', u'id': u'hypervisor.local', u'out': u'nested'}}
[...]
```
trace formatted:
```
The minion function caused an exception: 

Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1606, in _thread_return
    return_data = minion_instance.executors[fname](opts, data, func, args, kwargs)
  File "/usr/lib/python2.7/site-packages/salt/executors/direct_call.py", line 12, in execute
    return func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/modules/virt.py", line 767, in init
    __salt__[seed_cmd](
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1143, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/lib/python2.7/site-packages/salt/utils/lazy.py", line 99, in __getitem__
    if self._load(key):
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1621, in _load
    raise KeyError(\'The key must be a string.\')
KeyError: (u"The key must be a string.")
```

**Comment by rallytime** [comment]
This should be fixed with #48731
